src/transport_performance/population/rasterpop.py

```diff
@@ -16,6 +16,12 @@
 from shapely.geometry.polygon import Polygon
 from matplotlib import colormaps
 from cartopy.mpl.geoaxes import GeoAxes
+from transport_performance.utils.defence import (
+    _is_expected_filetype,
+    _type_defence,
+    _handle_path_like,
+    _check_parent_dir_exists,
+)
 
 
 class RasterPop:
```

Comment:
For the private plotting functions in `RasterPop`, type defences should be added.  
These functions include:  

- `._plot_folium()`
- `._plot_cartopy()`
- `._plot_matplotlib()`  

The rationale for this is that while they are private functions, kwargs can be passed to then through _.plot()_. Because of this, type defences should be included.

Reply:
Good point - I was toying with this and went on the side of maintainability. I was never fully convinced by my decision, so happy to do this. I'll work on it now since I have some time. In terms of unit tests, it's no so much about testing the methods of the functions inside `defence.py`, but more that unit tests are needed so that we can ensure the right types are being passed to the plotting functions, and that the correct errors are raised if it's not. Would you agree? If so, I'll add in respective unit tests too.

Reply:
Unit tests are entirely optional here in my opinion as there are tests for the defensive functions, meaning that we know they will work regardless. It's entirely up to you though!

Reply:
What are your thoughts on unit testing that the expected errors are raised if an incorrect type was passed as a kwarg? Not so much to test the `defence.py` function, but that the right type checking and error messages are being raised.

Reply:
This should now be closed out by the commits f0ec246, 8b2700f, ca2cc5b and af1a0b0. It's made the diff much larger, but these are primarily pytest parameterised configs for the unit tests being formatted over multiple line.

Let me know if these changes close out this conversation.